components/web_discovery/browser/server_config_loader.cc

```diff
@@ -0,0 +1,446 @@
+#include "base/threading/thread_restrictions.h"
+#include "base/values.h"
+#include "brave/components/web_discovery/browser/pref_names.h"
+#include "brave/components/web_discovery/browser/util.h"
+#include "components/prefs/pref_service.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "third_party/zlib/google/compression_utils.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr base::TimeDelta kMinReloadInterval = base::Hours(1);
+constexpr base::TimeDelta kMaxReloadInterval = base::Hours(4);
+
+constexpr base::TimeDelta kPatternsMaxAge = base::Hours(2);
+constexpr base::TimeDelta kPatternsRequestLatestDelay = base::Minutes(3);
+constexpr base::TimeDelta kPatternsRequestInitDelay = base::Seconds(15);
+
+constexpr size_t kPatternsMaxFileSize = 128000;
+
+constexpr net::NetworkTrafficAnnotationTag kNetworkTrafficAnnotation =
+    net::DefineNetworkTrafficAnnotation("wdp_config", R"(
+    semantics {
+      sender: "Brave Web Discovery Server Configuration Fetch"
+      description:
+        "Requests server configuration needed to send Web Discovery "
+        "measurements to Brave servers."
+      trigger:
+        "Requests are automatically sent at intervals "
+        "while Brave is running."
+      data: "Configuration attributes"
+      destination: WEBSITE
+    }
+    policy {
+      cookies_allowed: NO
+      setting:
+        "Users can opt-in or out via brave://settings/search"
+    })");
+
+constexpr char kGroupPubKeysFieldName[] = "groupPubKeys";
+constexpr char kPubKeysFieldName[] = "pubKeys";
+constexpr char kMinVersionFieldName[] = "minVersion";
+constexpr char kKeysFieldName[] = "keys";
+constexpr char kLimitFieldName[] = "limit";
+constexpr char kPeriodFieldName[] = "period";
+constexpr char kSourceMapFieldName[] = "sourceMap";
+constexpr char kSourceMapActionsFieldName[] = "actions";
+constexpr char kLocationFieldName[] = "location";
+
+constexpr char kCollectorConfigPathWithFields[] =
+    "/config?fields=minVersion,groupPubKeys,pubKeys,sourceMap";
+constexpr char kQuorumConfigPath[] = "/config";
+constexpr char kPatternsFilename[] = "wdp_patterns.json";
+
+constexpr char kOmittedLocationValue[] = "--";
+constexpr auto kAllowedReportLocations =
+    base::MakeFixedFlatSet<std::string_view>(
+        {"ar", "at", "au", "be", "br", "ca", "ch", "cn", "cz", "de",
+         "dk", "es", "fi", "fr", "gb", "gr", "hu", "in", "it", "jp",
+         "mx", "nl", "no", "pl", "ro", "ru", "se", "ua", "us"});
+
+KeyMap ParseKeys(const base::Value::Dict& encoded_keys) {
+  KeyMap map;
+  for (const auto [date, key_b64] : encoded_keys) {
+    std::vector<uint8_t> decoded_data;
+    // Decode to check for valid base64
+    if (!base::Base64Decode(key_b64.GetString())) {
+      continue;
```

Comment:
eh? why not store the decoded result?

Reply:
i did this to avoid having to re-encode the the key for the anonymous credential challenge (which happens in the `CredentialManager` during the join process, after config loading). however, it doesn't really matter since we also process the decoded data during the join process. no matter what, we'd be executing a double-call to either `Base64Decode` or `Base64Encode`. changed to store decoded data, since that's slightly cheaper and makes more sense.